src/NSubstitute/Proxies/CastleDynamicProxy/CastleDynamicProxyFactory.cs

```diff
@@ -46,12 +46,19 @@ public object GenerateProxy(ICallRouter callRouter, Type typeToProxy, Type[] add
         }
 
         /// <summary>
-        /// Allows to dynamically create a type in runtime. Returns an instance of <see cref="TypeBuilder"/>,
-        /// so type could be customized and built later.
+        /// Allows to dynamically create a type in runtime.
+        /// Use the <paramref name="typeBuildCallback"/> callback to define and build the type.
         /// </summary>
-        public TypeBuilder DefineDynamicType(string typeName, TypeAttributes flags)
+        /// <param name="typeBuildCallback">Callback used to construct the type.</param>
+        /// <returns>The result returned by <paramref name="typeBuildCallback"/> callback.</returns>
+        public Type DefineDynamicType(Func<ModuleBuilder, Type> typeBuildCallback)
         {
-            return _proxyGenerator.ProxyBuilder.ModuleScope.DefineType(true, typeName, flags);
+            var moduleBuilder = _proxyGenerator.ProxyBuilder.ModuleScope.ObtainDynamicModuleWithStrongName();
+
+            using (_proxyGenerator.ProxyBuilder.ModuleScope.Lock.ForWriting())
```

Comment:
@stakx Here is the question to you as an author of [this](https://github.com/castleproject/Core/pull/391) improvement. Currently I'm defining a dynamic interface to make a delegate proxy later. As a part of this work I need to define an internal `IsReadOnly` attribute type inside the dynamic assembly to later mark the `in` parameters with it (see `DefineIsReadOnlyAttributeIfNeeded()` method around). I would like to somehow guarantee that attribute type is created only once.

Given that you deprecated this lock, do you see any other ways to have such a guarantee?

Thanks for the help in advance! 👍

Reply:
@zvirja, thanks for bringing this to my attention. You're right that `ModuleBuilder`'s API now doesn't offer a good way to do thread-safe type creation now. The only possibility that I can see is ensuring in your library that your `ProxyGenerator` instance won't get used at all while you're generating that attribute type... i.e. having a separate lock of your own.

I've opened https://github.com/castleproject/Core/issues/399 to see what can be done at DynamicProxy's end.

Reply:
> The only possibility that I can see is ensuring in your library that your ProxyGenerator instance won't get used at all while you're generating that attribute type... i.e. having a separate lock of your own.

Unfortunately, it's very unsafe to proceed with such approach. It could happen that in future versions of `Castle.Core` it will also start to define the attribute. In this case there is a possibility of the race condition between my code and `Castle.Core`. Rather, we should use the same shared synchronization and play safe.

> I've opened castleproject/Core#399 to see what can be done at DynamicProxy's end. 

Thanks 👍 Let's try to proceed with that one and see how it goes.